events/api/src/main/java/org/projectnessie/events/api/Event.java

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.api;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.immutables.value.Value;
+import org.projectnessie.events.api.json.EventTypeIdResolver;
+
+/**
+ * Base interface for all events produced by Nessie.
+ *
+ * @see CommitEvent
+ * @see MergeEvent
+ * @see TransplantEvent
+ * @see ReferenceCreatedEvent
+ * @see ReferenceUpdatedEvent
+ * @see ReferenceDeletedEvent
+ * @see ContentStoredEvent
+ * @see ContentRemovedEvent
+ */
+@JsonSubTypes({
+  @JsonSubTypes.Type(name = "COMMIT", value = ImmutableCommitEvent.class),
+  @JsonSubTypes.Type(name = "MERGE", value = ImmutableMergeEvent.class),
+  @JsonSubTypes.Type(name = "TRANSPLANT", value = ImmutableTransplantEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_CREATED", value = ImmutableReferenceCreatedEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_UPDATED", value = ImmutableReferenceUpdatedEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_DELETED", value = ImmutableReferenceDeletedEvent.class),
+  @JsonSubTypes.Type(name = "CONTENT_STORED", value = ImmutableContentStoredEvent.class),
+  @JsonSubTypes.Type(name = "CONTENT_REMOVED", value = ImmutableContentRemovedEvent.class),
+  @JsonSubTypes.Type(name = "CUSTOM", value = ImmutableCustomEvent.class),
+})
+@JsonTypeIdResolver(EventTypeIdResolver.class)
+@JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, property = "type", visible = true)
+public interface Event {
+
+  /** The type of the event. */
+  EventType getType();
+
+  /**
+   * The id of the event. The UUID is generated by the event emitter. No assumptions should be made
+   * about the version of the UUID.
+   */
+  UUID getId();
+
+  /**
+   * The id of the repository. This is configured on a per-instance basis.
+   *
+   * <p>See configuration option: {@code nessie.version.store.advanced.repository-id}.
+   */
+  String getRepositoryId();
+
+  /**
+   * The time the event was created. The time is generated by the event emitter and is based on its
+   * internal wall clock time.
+   */
+  Instant getCreatedAt();
+
+  /**
+   * The user that created the event. The user is set by the event emitter and is based on the
+   * authenticated user that performed the action that caused the event to be emitted.
+   *
+   * <p>If authentication is disabled, this will be empty.
+   */
+  Optional<String> getCreatedBy();
+
+  /**
+   * A distinctive identifier for the entity that emitted the event. This is configured on a per-
+   * instance basis and usually contains the host name and the deployed Nessie version.
```

Comment:
> Why would we want to expose it in events?

I understand the security concern, but events are more secure than API cals. Do you think we should refrain from indicating the version?

> Would it make sense to use specVersion in events too?

Yes, I think so. I will make sure we add that – but keep reading :-)

> would it be preferable to separate the version/spec part from the actual instance/host ID?

Keeping in mind that this will be sent along with _every_ event, I think it's important to keep this lightweight. I suggest the following:

1. Keep this `getSentBy` attribute as just the server hostname, nothing more.
2. In the events SPI package (not reviewable yet but already implemented):
    1. leverage the `onSubscribe(EventSubscription)` method that is invoked once at startup for each subscriber;
    2. use the `EventSubscription` type to communicate to each subscriber the full server configuration, including but not limited to: server version, server spec version, default branch, etc. We can even pass the whole `NessieConfiguration` object, and maybe even more: the whole Quarkus configuration as a `Map<String,String>`. 
    3. If they wish, subscribers can store this info and augment the received events with all the info they need, e.g. when using Kafka, this can be easily done by setting headers on the `ProducerRecord`.

Wdyt?

In fact, if we go this route, we should even remove `repositoryId` from the Event interface as it is a global setting too.

Reply:
> we should even remove repositoryId from the Event interface

`repositoryId` would be needed for multi-tenancy.

> Would it make sense to use specVersion in events too?

IMO yes. A _spec_ version is fine. Not a fan of a particular server version, because then you have to know the exact _release_ version that introduces a particular change. A _spec_ version is IMO a better "grouping". It doesn't need to be the same spec version that we expose via `NessieConfiguration` though.

Why do we need the server / pod hostname? What's the use case for consumers?

Subscribers might need some configuration (could think of retry limits or so). But not the full configuration, which does include secrets. Let's discuss on the SPI PR.

Reply:
> repositoryId would be needed for multi-tenancy.

OK, let's keep it.

> Why do we need the server / pod hostname? What's the use case for consumers?

I thought it is important for consumers to know who / what created the event.

That said, server / pod hostname is also global to the Nessie instance, so we could remove `getSentBy` completely from the Event body and pass it to consumers using the `onSubscribe(EventSubscription)` method described above.

To get the ball rolling, I suggest that we remove `getSentBy` completely for now, and re-evaluate the best way to communicate this information + spec version to consumers, in the SPI PR.

Reply:
Done in f00e395.

Reply:
SGTM :+1: